### PR TITLE
Port marian to WASM for inference

### PR DESCRIPTION
**Pull request by abhi-agg**
### Description

This PR enables capability to build and use marian as a WASM module for doing inference.
It is related to issues: https://github.com/browsermt/marian-dev/issues/23

List of changes:
- Enable building only decoder (only for inference)
- Added onnxjs submodule (to provide a wasm compatible SGEMM routine)
- Updates intgemm submodule to use fast Integer 8-bit inference + simd shuffle pattern on WASM
- Updates SentencePiece submodule (to build stuff that is required just for inference)
- Build scripts (local machine and on Docker)
- Github workflow for WASM builds

Added dependencies: onnxjs submodule

### How to test
To run and test, follow the [README](https://github.com/browsermt/marian-dev/tree/wasm-merge-main/wasm)

### Checklist

- [x] I have tested the code manually
- [x] I have run regression tests
- [x] I have read and followed CONTRIBUTING.md
- [x] I have updated CHANGELOG.md

### CODE REVIEW

File: src/common/cli_helper.cpp [hunk 1]
```diff
@@ -15,6 +15,9 @@ void makeAbsolutePaths(YAML::Node& config,
       return nodePath;
 
     // replace relative path w.r.t. config directory
+  #if WITHOUT_EXCEPTIONS
+      return canonical(filesystem::Path{nodePath}, configDir).string();
```
Comment (ugermann) [inline src/common/cli_helper.cpp:19]: So what happens if the file does not exist? No error handling?

File: src/tensors/cpu/prod.cpp [hunk 2]
```diff
@@ -195,9 +197,10 @@ void ProdBatched(marian::Tensor C,
                  bool transB,
                  float beta,
                  float scalar) {
-  if (C->getBackend()->isLegacyBatchedGemm()) {
+  //if (C->getBackend()->isLegacyBatchedGemm()) {
     ProdBatchedOld(C, allocator, A, B, transA, transB, beta, scalar);
-  }
+  //}
+  /*
```
Comment (ugermann) [inline src/tensors/cpu/prod.cpp:203]: ... and below: Clean up please before submitting pull requests.

**Review by ugermann — CHANGES_REQUESTED** [review]
Before submitting pull requests please make sure that 
- all changes are properly documented in CHANGELOG.md 
- all regression tests pass
- all files with changes have clutter / commented-out code removed (or contain a convincing explanation why the commented-out code should be kept around

Without at least these three conditions having been met, chances of approval are very slim.

File: CMakeLists.txt [hunk 3]
```diff
@@ -197,24 +232,68 @@ else(MSVC)
...
-  set(CMAKE_C_FLAGS_RELWITHDEBINFO  "${CMAKE_C_FLAGS_RELEASE}")
-  set(CMAKE_C_FLAGS_PROFILE         "${CMAKE_C_FLAGS_RELEASE} -pg")
-  set(CMAKE_C_FLAGS_PROFGEN         "${CMAKE_C_FLAGS_RELEASE} -fprofile-generate -fprofile-correction")
-  set(CMAKE_C_FLAGS_PROFUSE         "${CMAKE_C_FLAGS_RELEASE} -fprofile-use -fprofile-correction")
+  if(COMPILE_WASM)
+    # Setting USE_SSE2 definition to enable SSE2 specific code in "3rd_party/sse_mathfun.h" for wasm builds
+    add_compile_definitions(USE_SSE2)
+    # Add compile definition for wasm builds
+    add_compile_definitions(WASM)
+    if(0)
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply CMakeLists.txt:240]: I will remove the dead code from here 👍

File: src/3rd_party/CMakeLists.txt [hunk 4]
```diff
@@ -78,18 +85,31 @@ if(USE_SENTENCEPIECE)
...
+  endif()
 
   if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    if(COMPILE_DECODER_ONLY)
+      set_property(TARGET sentencepiece-static APPEND_STRING PROPERTY COMPILE_FLAGS " -Wno-tautological-compare -Wno-unused")
+      if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 9.0)
+        set_property(TARGET sentencepiece-static APPEND_STRING PROPERTY COMPILE_FLAGS " -Wno-range-loop-construct")
+      endif()
+      # get_property(SENTENCEPIECE_COMPILE_FLAGS TARGET ${t} PROPERTY COMPILE_FLAGS)
+      # message("-- SENTENCPIECE: compile flags for target ${t}: ${SENTENCEPIECE_COMPILE_FLAGS}")
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/3rd_party/CMakeLists.txt:111]: Will do.

File: src/3rd_party/CMakeLists.txt [hunk 5]
```diff
@@ -123,13 +150,15 @@ include_directories(./CLI)
 include_directories(./pathie-cpp/include)
 
 if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
-  #set_target_properties(SQLiteCpp PROPERTIES COMPILE_FLAGS
-  set_property(TARGET SQLiteCpp APPEND_STRING PROPERTY COMPILE_FLAGS
-    " -Wno-parentheses-equality -Wno-unused-value")
-  if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 9.0)
+  if(NOT COMPILE_DECODER_ONLY)
+    #set_target_properties(SQLiteCpp PROPERTIES COMPILE_FLAGS
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/3rd_party/CMakeLists.txt:154]: Will do.

File: src/tensors/cpu/prod.cpp (hunk 2 above)
Comment (abhi-agg) [reply src/tensors/cpu/prod.cpp:203]: I am fine with cleaning up this file. However, I would like to have @XapaJIaMnu's opinion.
@XapaJIaMnu As I did this change [taking your code as reference](https://github.com/browsermt/marian-dev/pull/8), I would like to know if I can remove the dead code from here 👍

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -74,12 +78,12 @@ set(MARIAN_SOURCES
   layers/generic.cpp
   layers/loss.cpp
   layers/weight.cpp
-  layers/lsh.cpp
+  #layers/lsh.cpp
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/CMakeLists.txt]: Will do.

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -74,12 +78,12 @@ set(MARIAN_SOURCES
...
 
   rnn/cells.cpp
   rnn/attention.cpp
 
-  optimizers/quantizer.cpp
+  #optimizers/quantizer.cpp
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/CMakeLists.txt]: Will do.

File: src/CMakeLists.txt [hunk 8]
```diff
@@ -98,24 +102,51 @@ set(MARIAN_SOURCES
...
-  training/graph_group.cpp
-  training/graph_group_singleton.cpp
-  training/validator.cpp
-  training/communicator.cpp
+  #training/graph_group_async.cpp
+  #training/graph_group_sync.cpp
+  #training/graph_group.cpp
+  #training/graph_group_singleton.cpp
+  #training/validator.cpp
+  #training/communicator.cpp
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/CMakeLists.txt]: Will do.

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -98,24 +102,51 @@ set(MARIAN_SOURCES
...
 
   # this is only compiled to catch build errors, but not linked
-  microsoft/quicksand.cpp
-  microsoft/cosmos.cpp
+  #microsoft/quicksand.cpp
+  #microsoft/cosmos.cpp
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/CMakeLists.txt]: Will do.

File: src/CMakeLists.txt [hunk 10]
```diff
@@ -98,24 +102,51 @@ set(MARIAN_SOURCES
...
 
   $<TARGET_OBJECTS:libyaml-cpp>
-  $<TARGET_OBJECTS:SQLiteCpp>
+  #$<TARGET_OBJECTS:SQLiteCpp>
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/CMakeLists.txt]: Done.

File: src/CMakeLists.txt [hunk 11]
```diff
@@ -98,24 +102,51 @@ set(MARIAN_SOURCES
...
   $<TARGET_OBJECTS:pathie-cpp>
-  $<TARGET_OBJECTS:zlib>
-  $<TARGET_OBJECTS:faiss>
+  #$<TARGET_OBJECTS:zlib>
+  #$<TARGET_OBJECTS:faiss>
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/CMakeLists.txt]: Done.

File: docker/Makefile [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+head -n10 /repo/models/newstest2013.es > /repo/models/newstest2013.es.top10lines && \
+python3 /emsdk/upstream/emscripten/tools/file_packager.py ${BUILD_DIR}/model-files.data --preload ${MODELS_DIR} --js-output=${BUILD_DIR}/model-files.js'
+
+# Run WASM-without-pthreads-compiled marian-decoder
+run-wasm-without-pthreads: BUILD_DIR = /repo/build-wasm-without-pthreads-docker
+run-wasm-without-pthreads:
+	${run_on_docker} -p 8000:8000 ${WASM_IMAGE} bash -c 'emrun --no_browser --port 8000 ${BUILD_DIR}'
+
+
+# # Compile full native version
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply docker/Makefile:104]: Done.

File: docker/wasm/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    autoconf \
+    autotools-dev \
+    automake \
+    autogen \
+    libtool && ln -s /usr/bin/libtoolize /usr/bin/libtool
+    #&& mkdir -p /usr/opt \
+    #&& cd /usr/opt
+    #&& git clone https://github.com/menduz/protobuf-wasm-lib
+
+# RUN cd /usr/opt/protobuf-wasm-lib \
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply docker/wasm/Dockerfile:26]: Done.

File: src/tensors/cpu/prod.cpp (hunk 2 above)
Comment (XapaJIaMnu) [reply src/tensors/cpu/prod.cpp:203]: What I did here is an example to show you how to remove MKL from the codebase (and replace it with oneDNN). Since then there were other changes on top of my example (Did we end up using Eigen as a BLAS library?). You can safely remove the commented out lines, as there is no `sgemmBatched` implementation with the BLAS provider.

Comment (abhi-agg) [reply src/tensors/cpu/prod.cpp:203]: @XapaJIaMnu Just to be sure, I can remove [this dead code](https://github.com/browsermt/marian-dev/blob/wasm/src/tensors/cpu/prod.cpp#L203-L280) as well?

File: src/3rd_party/threadpool.h [hunk 14]
```diff
@@ -53,29 +53,46 @@ class ThreadPool {
     ~ThreadPool();
 
     size_t getNumTasks() const {
+    #if USE_PTHREADS
       return tasks.size();
+    #else
+      ABORT("ThreadPool::getNumTasks() not supported");
+    #endif
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/3rd_party/threadpool.h]: Agreed. I reverted all the changes in this file and submitted a change. I will resolve all the conversations in this file 👍

File: src/tensors/cpu/prod.cpp (hunk 2 above)
Comment (abhi-agg) [reply src/tensors/cpu/prod.cpp:203]: Cool. As @XapaJIaMnu confirmed in the comments, I will clean this up.

File: CHANGELOG.md [hunk 15]
```diff
@@ -53,6 +53,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Added default "none" for option shuffle in BatchGenerator, so that it works in executables where shuffle is not an option.
 - Added a few missing header files in shortlist.h and beam_search.h.
 - Improved handling for receiving SIGTERM during training. By default, SIGTERM triggers 'save (now) and exit'. Prior to this fix, batch pre-fetching did not check for this sigal, potentially delaying exit considerably. It now pays attention to that. Also, the default behaviour of save-and-exit can now be disabled on the command line with --sigterm exit-immediately.
+- Fix the runtime failures for FASTOPT on 32-bit builds (wasm just happens to be 32-bit) because it uses hashing with an inconsistent mix of uint64_t and size_t.
```
Comment (XapaJIaMnu) [inline CHANGELOG.md:56]: We should include fix for binary model loading in the 32 bit. I realise this is my fault though

File: src/3rd_party/pathie-cpp/src/path.cpp [hunk 16]
```diff
@@ -935,6 +935,8 @@ Path Path::exe()
 
   std::string str = utf16_to_utf8(buf);
   return Path(str);
+#elif defined(DECODER_ONLY)
+  throw(std::runtime_error("Path::exe() not supported in decoder-only version of marian"));
```
[the remark this answers is not in the document]
Comment (XapaJIaMnu) [reply src/3rd_party/pathie-cpp/src/path.cpp:939]: I would also like to see explanation for this one

**Review by XapaJIaMnu — APPROVED** [review]
Left 2 comments.
Could you please answer all @ugermann's concerns about the disabled exceptions before merging.

File: src/3rd_party/pathie-cpp/src/path.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (kpu) [reply src/3rd_party/pathie-cpp/src/path.cpp:939]: This is finding the name of an executable from within the executable.  Normally by opening /proc or something (or just argv[0]...).  I can understand that the existing code breaks in wasm since it assumes /proc .  And I can understand why it's not implemented, as unnecessary.  
Also, in defense of @abhi-agg, I thought exceptions just mapped to process death when exceptions were disabled?

File: CHANGELOG.md (hunk 15 above)
Comment (abhi-agg) [reply CHANGELOG.md:56]: @XapaJIaMnu Probably, I didn't understand. We already included [your fix](https://github.com/browsermt/marian-dev/pull/24/commits/e8ca9eacb253023235db0c83516d08f5959bd338) that solves this problem. Right? Do you mean just updating the CHANGELOG.md?

File: src/common/file_stream.cpp [hunk 17]
```diff
@@ -45,8 +47,12 @@ InputFileStream::InputFileStream(const std::string &file)
 
   // insert .gz decompression
   if(marian::utils::endsWith(file, ".gz")) {
+#if defined(DECODER_ONLY)
+    ABORT(".gz file decompression not supported in decoder-only build of Marian: {}", file);
```
[the remark this answers is not in the document]
Comment (ugermann) [reply src/common/file_stream.cpp:51]: That was my point, of sorts. DECODER_ONLY should not be synonymous with "This is a wasm compile". My suggestion would be to use cmake dependent options to set specific options that enable or disable certain features depending on the platform environment. If wasm compiles imply that zlib is not available, I'm fine with that, but

1. options should do what they say (e.g., NO_ZLIB)
2. they should be implemented via dependent options, so that we don't need a whole slew of -DWHATEVER={on|off} when running cmake.

So if the cmake determines that this is a wasm compile (e.g. by looking at the compiler executable name) and sets a slew of options (that do what they say), I'm happy with this. However, I can imagine a scenario (e.g. a translation server) where I don't want to compile and link a pile of executables that I won't need in the end, but U still want to be able to decompress .gz files on the fly.

**Comment by ugermann** [comment]
> Could you please answer all @ugermann's concerns about the disabled exceptions before merging.

Also, all regression tests should pass before a merge into master, and all changes need to be documented in CHANGELOG.md. Otherwise, there's no chance in hell we'll ever get this past the upstream code masters for an upstream merge.

File: src/common/file_stream.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/common/file_stream.cpp:51]: > options should do what they say (e.g., NO_ZLIB)
> they should be implemented via dependent options, so that we don't need a whole slew of -DWHATEVER={on|off} when running cmake.

@ugermann I agree with you and I will make changes 👍

[the remark this answers is not in the document]
Comment (ugermann) [reply src/common/file_stream.cpp:51]: With an eye to consistency of naming of options used, I'd actually suggest to use USE_ZLIB rather than NO_ZLIB. It's easier to remember them that way.

File: src/3rd_party/cnpy/cnpy.h [hunk 18]
```diff
@@ -133,6 +135,9 @@ namespace cnpy {
 
     template<typename T> void npz_save(std::string zipname, std::string fname, const T* data, const unsigned int* shape, const unsigned int ndims, std::string mode = "w")
     {
+#if defined(DECODER_ONLY)
+            throw std::runtime_error("npz_save: Platform not supported");
+#else
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/3rd_party/cnpy/cnpy.h:140]: I will change the `DECODER_ONLY` preprocessor to `WASM` that gets defined whenever we compile for WASM platform.

Compiling it without exceptions will catch the exceptions on JS side 👍

File: src/3rd_party/cnpy/cnpy.h [hunk 19]
```diff
@@ -265,6 +271,9 @@ namespace cnpy {
     static inline
     void npz_save(std::string zipname, const std::vector<NpzItem>& items)
     {
+#if defined(DECODER_ONLY)
+        throw std::runtime_error("npz_save: Platform not supported");
+#else
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/3rd_party/cnpy/cnpy.h:276]: Same reply as above

File: src/3rd_party/pathie-cpp/src/path.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/3rd_party/pathie-cpp/src/path.cpp:939]: I will change it so that this gets activated only if WASM compilation is enabled.

File: src/common/file_stream.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/common/file_stream.cpp]: I changed it to WASM as .gz compression is unsupported only for wasm builds. 👍

File: CHANGELOG.md (hunk 15 above)
Comment (abhi-agg) [reply CHANGELOG.md:58]: Updated the changelog to reflect the things added/fixed/changed by this PR 👍

File: docker/Makefile [hunk 20]
```diff
@@ -0,0 +1,240 @@
+# -*- mode: makefile-gmake; indent-tabs-mode: true; tab-width: 4 -*-
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply docker/Makefile:1]: Yes. It helps in reproducing the builds in a platform independent way. So it is helpful 👍

File: src/common/cli_helper.cpp (hunk 1 above)
Comment (abhi-agg) [reply src/common/cli_helper.cpp]: This comment is not valid anymore as I reverted this change 👍

File: src/3rd_party/CLI/App.hpp [hunk 21]
```diff
@@ -29,12 +29,16 @@
 namespace CLI {
 
 #ifndef CLI11_PARSE
+#if WITHOUT_EXCEPTIONS
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply src/3rd_party/CLI/App.hpp]: This comment is not valid anymore as I reverted this change 👍

**Comment by abhi-agg** [comment]
@ugermann @kpu @XapaJIaMnu I have made lot of improvements, answered most of your concerns in the comments and added commits to resolve them. I just need to run regression tests now. It would be great if you can review again. It looks in much better shape now 👍

**Comment by abhi-agg** [comment]
I have rebased this PR on `master` and ran the regression tests on:

1. Latest `master` ([commit](b08b26b5cfb9c6498ef48c1ee253f5bd5f788236))
2. On this PR

I am getting the **exact same status** for both of them which is following:
```
Failed:
  - tests/examples/unit-tests/test_unit_tests.sh
  - tests/scorer/scores/test_scores_cpu.sh  
```

1st test is failing because I have CUDA-11 and `cusparseSgemmi is not available in CUDA VERSION >= 11` (see the logs below). I believe this will pass for CUDA-10.

2nd test failure is already [known](https://github.com/browsermt/marian-dev/pull/16).

For the sake of completeness, I am sharing the failure logs for the 1st test case 👍 

<details>
 <summary>Failure logs for test_unit_tests.sh </summary>
 
 ```
 + set -e
+ cd /data/rw/pit/src/build-original
+ env CTEST_OUTPUT_ON_FAILURE=1 ctest --force-new-ctest-process
Test project /data/rw/pit/src/build-original
    Start 1: graph_tests
1/6 Test #1: graph_tests ......................   Passed    0.36 sec
    Start 2: operator_tests
2/6 Test #2: operator_tests ...................***Exception: Child aborted  0.99 sec
[2021-03-22 17:28:32] Error: cusparseSgemmi is not available in CUDA VERSION >= 11.
[2021-03-22 17:28:32] Error: Aborted from cusparseStatus_t marian::gpu::cusparseSgemmiEx(cusparseHandle_t, int, int, int, int, const float*, const float*, int, const float*, const int*, const int*, const float*, float*, int) in /data/rw/pit/src/src/tensors/gpu/prod.cpp:367

[CALL STACK]
[0x55cb46eed9c5]    marian::gpu::  CSRProd  (IntrusivePtr<marian::TensorBase>,  std::shared_ptr<marian::Allocator>,  IntrusivePtr<marian::TensorBase> const&,  IntrusivePtr<marian::TensorBase> const&,  IntrusivePtr<marian::TensorBase> const&,  IntrusivePtr<marian::TensorBase> const&,  bool,  bool,  float) + 0x1d65
[0x55cb4507057d]                                                       + 0x40d57d
[0x55cb450a1fa7]    std::_Function_handler<void (),marian::CSRDotNodeOp::forwardOps()::{lambda()#1}>::  _M_invoke  (std::_Any_data const&) + 0x207
[0x55cb4514f6d1]    marian::Node::  forward  ()                        + 0x211
[0x55cb45062d3b]    marian::ExpressionGraph::  forward  (std::__cxx11::list<IntrusivePtr<marian::Chainable<IntrusivePtr<marian::TensorBase>>>,std::allocator<IntrusivePtr<marian::Chainable<IntrusivePtr<marian::TensorBase>>>>>&,  bool) + 0x22b
[0x55cb450647bc]    marian::ExpressionGraph::  forwardNext  ()         + 0x23c
[0x55cb44f9a197]    void  tests  <float>(marian::DeviceType,  marian::Type) + 0xb5b7
[0x55cb44e9db8f]    Catch::RunContext::  invokeActiveTestCase  ()      + 0x2f
[0x55cb44ead18d]    Catch::RunContext::  runCurrentTest  (std::__cxx11::basic_string<char,std::char_traits<char>,std::allocator<char>>&,  std::__cxx11::basic_string<char,std::char_traits<char>,std::allocator<char>>&) + 0x47d
[0x55cb44ec2030]    Catch::RunContext::  runTest  (Catch::TestCase const&) + 0x1a0
[0x55cb44ec8b5c]    Catch::Session::  runInternal  ()                  + 0xa3c
[0x55cb44ec9ada]    Catch::Session::  run  ()                          + 0x1a
[0x55cb44e69186]    main                                               + 0x86
[0x7f13b6a18b97]    __libc_start_main                                  + 0xe7
[0x55cb44e9539a]    _start                                             + 0x2a


~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
run_operator_tests is a Catch v2.10.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
Expression graph supports basic math operations (gpu)
  csr-dot product
-------------------------------------------------------------------------------
/data/rw/pit/src/src/tests/units/operator_tests.cpp:456
...............................................................................

/data/rw/pit/src/src/tests/units/operator_tests.cpp:456: FAILED:
  {Unknown expression after the reported line}
due to a fatal error condition:
  SIGABRT - Abort (abnormal termination) signal

===============================================================================
test cases:  1 |  0 passed | 1 failed
assertions: 81 | 80 passed | 1 failed


    Start 3: rnn_tests
3/6 Test #3: rnn_tests ........................   Passed    0.79 sec
    Start 4: attention_tests
4/6 Test #4: attention_tests ..................   Passed    0.69 sec
    Start 5: fastopt_tests
5/6 Test #5: fastopt_tests ....................   Passed    0.02 sec
    Start 6: utils_tests
6/6 Test #6: utils_tests ......................   Passed    0.02 sec

83% tests passed, 1 tests failed out of 6

Total Test time (real) =   3.03 sec

The following tests FAILED:
          2 - operator_tests (Child aborted)
Errors while running CTest
  ```
</details>

**TL;DR: This PR doesn't introduce any additional regression test failures to current master.**

@ugermann @kpu Please review the PR as all the changes that you requested are in. I would like to merge it asap.

**Comment by jerinphilip** [comment]
> TL;DR: This PR doesn't introduce any additional regression test failures to current master.

It does not strictly suffice here. Plenty of (CMake) switches in this branch has previously led to failure in linkages (quantizer.cpp for example), from bergamot-translator switches. Someone trying to build shouldn't be able to mess with switches and get a non-working build. Normally Makefiles for separate architectures (Makefile, Makefile.mingw) etc are supposed to be loaded or unloaded as a whole, but this PR makes it goes through a lot many switches making it painful to review. I would personally be scared to branch off browsermt/marian-dev for development if this gets merged in. 

Your use case is only target WASM architecture (which you're testing and reporting), but the build logic change seems to generalize many other combinations - decoder only, no exceptions, etc etc. 

> I would like to merge it asap.

For all complaints of big PR elsewhere with even tinier set of changes, this is quite interesting.

File: CMakeLists.txt [hunk 22]
```diff
@@ -8,24 +8,63 @@ endif ()
...
 option(COMPILE_TESTS "Compile tests" OFF)
+option(COMPILE_WASM "Compile marian as WASM target" OFF)
+option(COMPILE_DECODER_ONLY "Compile marian-decoder only" OFF)
+option(COMPILE_WITH_PTHREADS "Compile with pthreads support" ON)
 option(USE_CCACHE "Use ccache compiler cache (https://ccache.dev)" OFF)
 option(USE_CUDNN "Use CUDNN library" OFF)
 option(USE_DOXYGEN "Build documentation with Doxygen" ON)
 option(USE_FBGEMM "Use FBGEMM" OFF)
 option(USE_MKL "Compile with MKL support" ON)
+option(USE_WASM_COMPATIBLE_BLAS "Compile with a WASM compatible blas for decoder only builds" OFF)
```
Comment (jerinphilip) [inline CMakeLists.txt:26]: I think you should expose only `COMPILE_WASM` or something similar outside as an option in this PR, and set/unset the remaining features under `if(COMPILE_WASM)` and else given the supplied GitHub build file only tests it for WASM (and not the others).

File: src/layers/generic.cpp [hunk 23]
```diff
@@ -221,9 +221,13 @@ namespace marian {
 
       // this option is only set in the decoder
       if(!lsh_ && options_->hasAndNotEmpty("output-approx-knn")) {
+#ifdef DECODER_ONLY
```
Comment (kpu) [inline src/layers/generic.cpp:224]: LSH is something that's really only useful for decoding.  And we might be replacing lexical shortlists with it in the future.  The `DECODER_ONLY` macro is the wrong one to use here.

Comment (abhi-agg) [reply src/layers/generic.cpp:224]: Got it. I will replace it with `COMPILE_WASM` macro.

@kpu Is there any other concern from your side on this PR? I would like to address all of them with one commit 👍

File: CMakeLists.txt (hunk 22 above)
Comment (abhi-agg) [reply CMakeLists.txt]: Done. All cmake options that are wasm specific are encapsulated under one cmake option (`COMPILE_WASM` in our case).

**Comment by abhi-agg** [comment]
> > TL;DR: This PR doesn't introduce any additional regression test failures to current master.
> 
> It does not strictly suffice here. Plenty of (CMake) switches in this branch has previously led to failure in linkages (quantizer.cpp for example), from bergamot-translator switches. Someone trying to build shouldn't be able to mess with switches and get a non-working build. Normally Makefiles for separate architectures (Makefile, Makefile.mingw) etc are supposed to be loaded or unloaded as a whole, but this PR makes it goes through a lot many switches making it painful to review. I would personally be scared to branch off browsermt/marian-dev for development if this gets merged in (due to the sheer complexity it induces on the build files which will be tricky to edit after).
> 
> Your use case is only target WASM architecture (which you're testing and reporting), but the build logic change seems to generalize many other combinations - decoder only, no exceptions, etc etc.
> 

@jerinphilip The recent commits in this PR addresses your concerns 👍

File: CMakeLists.txt [hunk 24]
```diff
@@ -26,6 +27,49 @@ option(USE_NCCL "Use NCCL library" ON)
 option(USE_SENTENCEPIECE "Download and compile SentencePiece" ON)
 option(USE_STATIC_LIBS "Link statically against non-system libs" OFF)
 option(GENERATE_MARIAN_INSTALL_TARGETS "Generate Marian install targets (requires CMake 3.12+)" OFF)
+option(M32_BINARIES "Generate 32bit binaries even when building outside outside of WASM. Useful for" OFF)
```
Comment (kpu) [inline CMakeLists.txt:30]: Unfinished sentence?

Comment (abhi-agg) [reply CMakeLists.txt:30]: This change came from @XapaJIaMnu.
@XapaJIaMnu Do you have any preference on what to write here?

File: CMakeLists.txt [hunk 25]
```diff
@@ -26,6 +27,49 @@ option(USE_NCCL "Use NCCL library" ON)
...
+option(COMPILE_WASM "Compile customized marian for WASM target" OFF)
+
+# cmake options that are dependent on COMPILE_WASM cmake option
+CMAKE_DEPENDENT_OPTION(COMPILE_DECODER_ONLY "Compile marian-decoder only" ON
+                       "COMPILE_WASM" OFF)
+CMAKE_DEPENDENT_OPTION(COMPILE_WITH_PTHREADS "Compile with pthreads support" OFF
```
Comment (kpu) [inline CMakeLists.txt:36]: To the extent that there is a style with marian-dev's options, `COMPILE_` controls whether a build product exists like `COMPILE_SERVER` while `USE_` controls whether a particular functionality/library is supported like `USE_SENTENCEPIECE`.  This feels more like USE_THREADS.

File: CMakeLists.txt [hunk 26]
```diff
@@ -435,29 +508,37 @@ if(USE_MPI)
...
-        include_directories(${BLAS_INCLUDE_DIR} ${CBLAS_INCLUDE_DIR})
-        set(EXT_LIBS ${EXT_LIBS} ${BLAS_LIBRARIES} ${CBLAS_LIBRARIES})
-        add_definitions(-DCOMPILE_CPU=1 -DBLAS_FOUND=1)
-      endif(CBLAS_FOUND)
-    endif(BLAS_FOUND)
-  endif(MKL_FOUND)
+    set(BLAS_VENDOR "ONNX-SGEMM")
+    add_compile_definitions(COMPILE_CPU  BLAS_FOUND  WASM_COMPATIBLE_BLAS)
+  else(USE_WASM_COMPATIBLE_BLAS)
+    set(EXT_LIBS ${EXT_LIBS} intgemm)
```
Comment (kpu) [inline CMakeLists.txt:520]: nit could have been outside the if statement

File: src/3rd_party/CMakeLists.txt [hunk 27]
```diff
@@ -78,18 +84,29 @@ if(USE_SENTENCEPIECE)
...
   include_directories(./sentencepiece)
-
+  if(NOT COMPILE_DECODER_ONLY)
   set_target_properties(spm_encode spm_decode spm_train spm_normalize spm_export_vocab
     PROPERTIES RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}")
+  endif()
 
   if (CMAKE_CXX_COMPILER_ID MATCHES "GNU" AND (CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL 8.0 OR CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 8.0))
+  if(COMPILE_DECODER_ONLY)
+    set_property(TARGET sentencepiece-static APPEND_STRING PROPERTY COMPILE_FLAGS " -Wno-class-memaccess")
```
Comment (kpu) [inline src/3rd_party/CMakeLists.txt:94]: Why does disabling warnings apply only if we're compiling only the decoder?

File: src/3rd_party/cnpy/cnpy.h [hunk 28]
```diff
@@ -5,7 +5,9 @@
 #ifndef LIBCNPY_H_
 #define LIBCNPY_H_
 
+#if !defined(WASM)
```
Comment (kpu) [inline src/3rd_party/cnpy/cnpy.h:8]: Fun fact: now that @XapaJIaMnu has binary working, you can get rid of cnpy entirely.

File: CMakeLists.txt (hunk 24 above)
Comment (XapaJIaMnu) [reply CMakeLists.txt:30]: Oh god, sorry, i have forgotten to finish that. It's supposed to be `Generate 32bit binaries even when building outside of WASM. Useful for testing some WASM specific functionality without the need for the compiling to WASM.` Could you change it?

File: src/3rd_party/CMakeLists.txt (hunk 27 above)
Comment (abhi-agg) [reply src/3rd_party/CMakeLists.txt:94]: Actually, I am removing this.

File: src/3rd_party/cnpy/cnpy.h (hunk 28 above)
Comment (abhi-agg) [reply src/3rd_party/cnpy/cnpy.h:8]: Yes, but for compilation we will still have to keep this ifdef. Right?

File: CMakeLists.txt (hunk 25 above)
Comment (abhi-agg) [reply CMakeLists.txt:36]: Fixing it in the next commit.

Comment (abhi-agg) [reply CMakeLists.txt]: Fixed.

File: docker/Makefile [hunk 29]
```diff
@@ -0,0 +1,240 @@
...
+
+make_cmd  = make -j3
+#make_cmd += VERBOSE=1
+
+# ... and running things on Docker
+docker_mounts  = ${PWD}/..:/repo
+docker_mounts += ${HOME}/.ccache:/.ccache
+run_on_docker  = docker run --rm
+run_on_docker += $(addprefix -v, ${docker_mounts})
+run_on_docker += ${INTERACTIVE_DOCKER_SESSION}
```
[the remark this answers is not in the document]
Comment (abhi-agg) [reply docker/Makefile]: @ugermann I build the docker version and it is not creating any issue of permission mapping for me. I ran the translation in the browser as well and that works too. 👍

**Comment by abhi-agg** [comment]
@kpu I made the final changes addressing all your concerns. A quick final review and approval will be much appreciated 👍

**Comment by abhi-agg** [comment]
@ugermann Could you please approve the PR as I already addressed all your review comments?

**Comment by abhi-agg** [comment]
> Before submitting pull requests please make sure that
> 
>     * all changes are properly documented in CHANGELOG.md
> 
>     * all regression tests pass
> 
>     * all files with changes have clutter / commented-out code removed (or contain a convincing explanation why the commented-out code should be kept around
> 
> 
> Without at least these three conditions having been met, chances of approval are very slim.

@ugermann Could you please approve the PR as I already addressed all your review comments?

**Review by ugermann — APPROVED** [review]

**Comment by ugermann** [comment]
I'd suggest a squashed merge.

**Comment by abhi-agg** [comment]
This PR is also connected to the https://github.com/browsermt/marian-dev/issues/5
